### PR TITLE
TransformerNetwork

### PR DESCRIPTION
**Pull request by emailweixu**
TransformerNetwork can be a drop-in replacement for LSTMEncodingNetwork

Optimize memory usage for TransformerBlock

### CODE REVIEW

File: alf/layers.py [hunk 1]
```diff
@@ -65,6 +65,58 @@ def forward(self, x):
...
+        if dim1 >= 0:
+            dim1 += 1
+        self._dim1 = dim1
+
+    def forward(self, x):
+        return x.transpose(self._dim0, self._dim1)
+
+
+class Permute(nn.Module):
+    """A layer that perform the transpose of channels."""
```
Comment (Haichao-Zhang) [inline alf/layers.py]: transpose -> permute

File: alf/layers.py [hunk 2]
```diff
@@ -65,6 +65,58 @@ def forward(self, x):
...
+        return x.to(self._dtype)
+
+
+class Transpose(nn.Module):
+    """A layer that perform the transpose of channels."""
+
+    def __init__(self, dim0=0, dim1=1):
+        """
+        Args:
+            dim0 (int): the first dimension to be transposed
```
Comment (Haichao-Zhang) [inline alf/layers.py]: Might be better to comment that ```dim``` does not include batch dimension.

File: alf/layers.py [hunk 3]
```diff
@@ -1134,33 +1213,44 @@ def __init__(self,
...
             num_heads (int): the number of attention heads
             memory_size (int): maximal allowed sequence length
-            d_k (int): Dimension of key, same as d_k in [1]. If None, use ``d_model``
-            d_v (int): Dimension of value, same as d_v in [1]. If None, use ``d_model``
-            relative_positional_encoding (bool): If True, use the relative positional
-                encoding proposed in [3]. If False, use absolute positional encoding.
+            d_k (int): Dimension of key, same as d_k in [1]. If None, use ``d_model // num_heads``
+            d_v (int): Dimension of value, same as d_v in [1]. If None, use ``d_model // num_heads``
+            d_ff (int): Diemension of the MLP, same as d_ff in [1]. If None, use ``4 * d_model``
+            positional_encoding (bool): One of ['none', 'abs', 'rel']. If 'none',
```
Comment (Haichao-Zhang) [inline alf/layers.py]: ```bool``` -> ```str```

File: alf/algorithms/merlin_algorithm.py [hunk 4]
```diff
@@ -27,7 +27,7 @@
 from alf.algorithms.algorithm import Algorithm
 from alf.algorithms.config import TrainerConfig
 from alf.algorithms.decoding_algorithm import DecodingAlgorithm
-from alf.algorithms.memory import MemoryWithUsage
+from alf.networks.memory import MemoryWithUsage
```
Comment (Haichao-Zhang) [inline alf/algorithms/merlin_algorithm.py]: we can move this line after line 33

File: alf/layers.py [hunk 5]
```diff
@@ -65,6 +65,58 @@ def forward(self, x):
...
+class Permute(nn.Module):
+    """A layer that perform the transpose of channels."""
+
+    def __init__(self, *dims):
+        """
+        Args:
+            *dims: The desired ordering of dimensions (not including batch dimension)
+        """
+        super().__init__()
+        dims = [1 + d for d in dims]
```
Comment (hnyu) [inline alf/layers.py:119]: Do we need to make sure that every dim is non-negative?

File: alf/networks/memory.py [hunk 6]
```diff
@@ -352,3 +352,97 @@ def from_states(self, states):
             self._memory, self._usage = states
             self._batch_size = self._memory.shape[0]
             self._built = True
+
+
+class FIFOMemory(Memory):
+    """A Simple FIFO memory.
+
+    Whwn new memory are written, the oldest memory slots are removed.
```
Comment (hnyu) [inline alf/networks/memory.py]: When new memory slots are ...

File: alf/networks/memory.py [hunk 7]
```diff
@@ -352,3 +352,97 @@ def from_states(self, states):
...
+        self._batch_size = batch_size
+        self._memory = torch.zeros(batch_size, self.size, self.dim)
+        self._current_size = torch.zeros(batch_size, dtype=torch.int64)
+        self._built = True
+
+    def write(self, content):
+        """Write content to memory.
+
+        Append the content to memory. If the memory is full, the oldest slot
+        removed.
```
Comment (hnyu) [inline alf/networks/memory.py]: will be removed

File: alf/networks/memory.py [hunk 8]
```diff
@@ -352,3 +352,97 @@ def from_states(self, states):
...
+
+        Args:
+            content (Tensor): shape should be [b, dim] or [b, k, dim]
```
Comment (hnyu) [inline alf/networks/memory.py]: should explain what 'k' means

File: alf/networks/memory.py [hunk 9]
```diff
@@ -352,3 +352,97 @@ def from_states(self, states):
...
+        User does not need to call this explictly. `read` and `write` will
+        automatically call this if the memory has not been built yet.
+
+        Note: Subsequent `write` and `read` must match this `batch_size`
+        Args:
+            batch_size (int): batch size of the model.
+        """
+        self._batch_size = batch_size
+        self._memory = torch.zeros(batch_size, self.size, self.dim)
+        self._current_size = torch.zeros(batch_size, dtype=torch.int64)
```
Comment (hnyu) [inline alf/networks/memory.py:393]: Seems that self._current_size can only be a scalar because it's the same across the batch

File: alf/algorithms/merlin_algorithm.py (hunk 4 above)
Comment (emailweixu) [reply alf/algorithms/merlin_algorithm.py]: moved

File: alf/networks/transformer_networks.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+    """A Network composed of Memory and TransformerBlock.
+
+    The following is the pseudocode for the computation:
+
+    .. code-block:: python
+
+        for i in range(num_prememory_layers):
+            core, inputs = T_i([core, inputs], [core, inputs])
+        for j in range(num_memory_layers):
+            core, inputs = TM_j([memory_j, core, inputs], [core, inputs])
```
Comment (hnyu) [inline alf/networks/transformer_networks.py:36]: This description is not accurate, because TransformerBlock only accepts tensors. Here it looks like lists.

File: alf/layers.py (hunk 2 above)
Comment (emailweixu) [reply alf/layers.py]: Added.

File: alf/layers.py (hunk 1 above)
Comment (emailweixu) [reply alf/layers.py]: Fixed

File: alf/networks/transformer_networks.py [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+import alf
+from alf.networks import PreprocessorNetwork
+from alf.networks.memory import FIFOMemory
+from alf.nest.utils import NestConcat
+
+
+@gin.configurable
+class TransformerNetwork(PreprocessorNetwork):
+    """A Network composed of Memory and TransformerBlock.
```
Comment (hnyu) [inline alf/networks/transformer_networks.py:27]: i think it's easy to confuse the memory of this network with the memory concept of a TransformerBlock. prememory layers still use TransformerBlock's memory, just not the memory created by this network. These two concepts can be better differentiated.

File: alf/layers.py (hunk 3 above)
Comment (emailweixu) [reply alf/layers.py]: fixed

File: alf/networks/transformer_networks.py [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+                 d_ff,
+                 centralized_memory=True,
+                 input_preprocessors=None,
+                 name="TransformerNetwork"):
+        """
+        Args:
+            input_tensor_spec (nested TensorSpec): the (nested) tensor spec of
+                the input.
+            memory_size (int): size of memory.
+            core_size (int): size of core (i.e. number of embeddings of core)
```
Comment (hnyu) [inline alf/networks/transformer_networks.py:74]: The concept of 'core' can be better explained.

File: alf/networks/transformer_networks.py [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+                continuous vector. The output_spec of each input preprocessor i
+                should be [input_size_i, d_model]. The result of all the preprocessors
+                will be concatenated as a Tensor of shape ``[batch_size, input_size, d_model]``,
+                where ``input_size = sum_i input_size_i``.
+
+        """
+        super().__init__(
+            input_tensor_spec,
+            input_preprocessors,
+            preprocessing_combiner=NestConcat(dim=-2),
```
Comment (hnyu) [inline alf/networks/transformer_networks.py]: Why it's -2? The input_tensor_spec argument doesn't say anything about the input format.

**Review by hnyu — CHANGES_REQUESTED** [review]

File: alf/networks/memory.py (hunk 9 above)
Comment (emailweixu) [reply alf/networks/memory.py:393]: Not necessarily. At the beginning of an episode, the state will be reset to zero. So different environment has different current_size

File: alf/networks/memory.py (hunk 7 above)
Comment (emailweixu) [reply alf/networks/memory.py]: Fixed

File: alf/networks/memory.py (hunk 8 above)
Comment (emailweixu) [reply alf/networks/memory.py]: Added

File: alf/networks/transformer_networks.py (hunk 10 above)
Comment (emailweixu) [reply alf/networks/transformer_networks.py:36]: This is just pseudocode. `[x, y]` just means concatenate x and y together, which I feel is understandable.

File: alf/networks/transformer_networks.py (hunk 11 above)
Comment (emailweixu) [reply alf/networks/transformer_networks.py:27]: For TransformerBlock, memory is just an argument. TransformerBlock itself does not have any memory.

File: alf/networks/transformer_networks.py (hunk 12 above)
Comment (emailweixu) [reply alf/networks/transformer_networks.py:74]: Added.

File: alf/networks/transformer_networks.py (hunk 13 above)
Comment (emailweixu) [reply alf/networks/transformer_networks.py]: I only explained this at ``input_preprocessors``. Now added for input_tensor_spec.

File: alf/layers.py (hunk 5 above)
Comment (emailweixu) [reply alf/layers.py:119]: assertion added.

**Review by Haichao-Zhang — DISMISSED** [review]
I have no further questions.

File: alf/networks/transformer_networks.py (hunk 11 above)
Comment (hnyu) [reply alf/networks/transformer_networks.py:27]: > For TransformerBlock, memory is just an argument. TransformerBlock itself does not have any memory.

Yeah that's the reason why it can get confusing because the first argument of line 34 is also called "memory" inside the TransformerBlock.

File: alf/networks/memory.py (hunk 9 above)
Comment (hnyu) [reply alf/networks/memory.py:393]: > Not necessarily. At the beginning of an episode, the state will be reset to zero. So different environment has different current_size

Would it be better to add a sentence saying that all memory classes in memory.py are episodic?

Comment (emailweixu) [reply alf/networks/memory.py:393]: Added.

File: alf/networks/transformer_networks.py (hunk 11 above)
Comment (emailweixu) [reply alf/networks/transformer_networks.py:27]: Added comments about this.

**Review by hnyu — DISMISSED** [review]

**Review by Haichao-Zhang — DISMISSED** [review]

**Review by hnyu — APPROVED** [review]

**Review by Haichao-Zhang — APPROVED** [review]
